runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -0,0 +1,1193 @@
+  if vim.tbl_isempty(line_diagnostics) then return end
+
+  for i, diagnostic in ipairs(line_diagnostics) do
+    local prefix = string.format("%d. ", i)
+    local hiname = M._get_floating_severity_highlight_name(diagnostic.severity)
+    assert(hiname, 'unknown severity: ' .. tostring(diagnostic.severity))
+
+    local message_lines = vim.split(diagnostic.message, '\n', true)
+    table.insert(lines, prefix..message_lines[1])
+    table.insert(highlights, {#prefix + 1, hiname})
+    for j = 2, #message_lines do
+      table.insert(lines, message_lines[j])
+      table.insert(highlights, {0, hiname})
+    end
+  end
+
+  local popup_bufnr, winnr = util.open_floating_preview(lines, 'plaintext')
+  for i, hi in ipairs(highlights) do
+    local prefixlen, hiname = unpack(hi)
+    -- Start highlight after the prefix
+    api.nvim_buf_add_highlight(popup_bufnr, -1, hiname, i-1, prefixlen, -1)
+  end
+
+  return popup_bufnr, winnr
+end
+
+local loclist_type_map = {
+  [DiagnosticSeverity.Error] = 'E',
+  [DiagnosticSeverity.Warning] = 'W',
+  [DiagnosticSeverity.Information] = 'I',
+  [DiagnosticSeverity.Hint] = 'I',
+}
+
+--- Sets the location list
+---@param opts table|nil Configuration table. Keys:
+---         - {open_loclist}: (boolean, default true)
+---             - Open loclist after set
+---         - {client_id}: (number)
+---             - If nil, will consider all clients attached to buffer.
+---         - {severity}: (DiagnosticSeverity)
+---             - Exclusive severity to consider. Overrides {severity_limit}
+---         - {severity_limit}: (DiagnosticSeverity)
+---             - Limit severity of diagnostics found. E.g. "Warning" means { "Error", "Warning" } will be valid.
+function M.set_loclist(opts)
+  opts = opts or {}
+
+  local open_loclist = if_nil(opts.open_loclist, true)
+
+  local bufnr = vim.api.nvim_get_current_buf()
+  local buffer_diags = M.get(bufnr, opts.client_id)
+
+  local severity = to_severity(opts.severity)
+  local severity_limit = to_severity(opts.severity_limit)
+
+  local items = {}
+  local insert_diag = function(diag)
+    if severity then
+      -- Handle missing severities
+      if not diag.severity then
+        return
+      end
+
+      if severity ~= diag.severity then
+        return
+      end
+    elseif severity_limit then
+      if not diag.severity then
+        return
+      end
+
+      if severity_limit < diag.severity then
+        return
+      end
+    end
+
+    local pos = diag.range.start
+    local row = pos.line
+    local col = util.character_offset(bufnr, row, pos.character)
+
+    local line = (api.nvim_buf_get_lines(bufnr, row, row + 1, false) or {""})[1]
+
+    table.insert(items, {
+      bufnr = bufnr,
+      lnum = row + 1,
+      col = col + 1,
+      text = line .. " | " .. diag.message,
+      type = loclist_type_map[diag.severity or DiagnosticSeverity.Error] or 'E',
+    })
+  end
+
+  for _, diag in ipairs(buffer_diags) do
+    insert_diag(diag)
+  end
```

Comment:
This is not sorted by default.
And a unsorted loclist breaks `labove` and `lbelow`.

Can you add sorting here (or earlier in the chain, not sure if it is relevant anywhere else)
```
table.sort(items, function(a,b) return a.lnum < b.lnum end)
```

Reply:
Thanks! And I added some tests for it too :)